airflow/www/utils.py

```diff
@@ -133,13 +133,25 @@ def get_mapped_summary(parent_instance, task_instances):
 
 
 def get_dag_run_conf(dag_run_conf: Any) -> tuple[str | None, bool]:
+    class DagRunConfEncoder(json.JSONEncoder):
+        def default(self, obj):
+            if isinstance(obj, bytes):
+                try:
+                    return obj.decode()
+                except Exception:
+                    return str(obj)
+            try:
+                return json.JSONEncoder.default(self, obj)
+            except Exception:
+                return str(obj)
```

Comment:
This method is only used to show dag run conf dictionary content in Airflow Webserver, but the same content is stored in the database as `PickleType` without any type conversion, and it read and deserialized when we call `dag_run.conf`.
As long as there are no problems with handling the different types in the conf dict, I think converting binary types to str can solve the UI problem.

Reply:
Another concern I have no idea how to add non-JSON Serializable objects into the `conf` by use public parts of Airflow and should we encourage the use of undocumented features

Reply:
Actually It's possible using the Airflow python API (the method [DAG.create_dagrun](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/dag/index.html#airflow.models.dag.DAG.create_dagrun)):
```python
dag_bag = DagBag(read_dags_from_db=True)
dag_bag.collect_dags_from_db()
dag = dag_bag.get_dag("dag_id")
dag.create_dagrun(
    run_id="run_id",
    run_type=DagRunType.MANUAL,
    execution_date=run_start_date,
    state=State.QUEUED,
    external_trigger=True,
    data_interval=(run_start_date, run_end_date),
    conf = {<str key>: <any value>}
)
```
And we use this method frequently in the [custom plugins](https://airflow.apache.org/docs/apache-airflow/stable/plugins.html). This part is public and documented.
Also we can update the run conf from a task, and the updated fields are accessible from the other tasks ([ex](https://github.com/apache/airflow/issues/28772))

Reply:
Well, `DagBag` it is definitely not a part of public interface and intend by use in plugins and by users as well as manually call `DAG.create_dagrun` outside of the tests, the only one operator directly use it is SubDagOperator, which deprecated and will be remove in Airflow 3

The general usage of the DAG is create DAG object and assign it by one of the [available way](https://airflow.apache.org/docs/apache-airflow/stable/concepts/dags.html#dag-assignment).

For run DAG user could use REST API/CLI/TriggerDagRunOperator/Scheduler, all other ways is not documented and definitely not recommended. I found same issue when TriggerDagRunOperator supported send non-json serialisable data to `conf`, and solution was ban this ability, see https://github.com/apache/airflow/issues/13414

Not all methods/functions/classes have `_` in the beginning but it doesn't mean that this methods could directly call by programmatic way. Best example is `execute` method of `BaseOperator` which only intend to use by Scheduler/Executor.

There is also PR https://github.com/apache/airflow/pull/28300 which prepare info about public Airflow interface, I would suggest clarifying about usage some part of airflow (DagBag, DAG.create_dagrun) in that PR because it might help to cover more grey zones.

Reply:
Just to comment, if you _have_ to stringify things for display, `repr` is better than `str`.

Reply:
I just checked and indeed this function is only used in `views.py`. However since there’s no way to tell whether it’s also used by user code somewhere (backward compatibility concerns), it’s probably better to make the function signature something like this:

```python
def get_dag_run_conf(
    dag_run_conf: Any,
    *,
    json_encoder: Type[json.JSONEncoder] = json.JSONEncoder,
) -> tuple[str | None, bool]:
    ...
```

and we can move `DagRunConfEncoder` to `vies.py` as a private class.